pydm/data_plugins/calc_plugin.py

```diff
@@ -0,0 +1,204 @@
+            c = pydm.PyDMChannel(ch, connection_slot=conn_cb,
+                                 value_slot=value_cb)
+            self._channels.append(c)
+
+    @property
+    def connected(self):
+        return all(self._connections)
+
+    def _connect(self):
+        self._values = [None]*len(self._channels)
+        self._connections = [False]*len(self._channels)
+        for ch in self._channels:
+            ch.connect()
+
+    def _disconnect(self):
+        for ch in self._channels:
+            ch.disconnect()
+
+    def _send_update(self, conn, value):
+        self.new_data_signal.emit({"connection": conn,
+                                   "value": value})
+
+    def run(self):
+        self._connect()
+
+        while True:
+            self._calculate.wait()
+            self._calculate.clear()
+            if self.isInterruptionRequested():
+                break
+            self.calculate_expression()
+        self._disconnect()
+
+    def callback_value(self, ch_index, value):
+        """
+        Callback executed when a channel receives a new value.
+
+        Parameters
+        ----------
+        ch_index : int
+            The channel index on the list for this rule.
+        value : any
+            The new value for this channel.
+
+        Returns
+        -------
+        None
+        """
+        self._values[ch_index] = value
+        if not self.connected:
+            self.warn_unconnected_channels()
+            return
+        self._calculate.set()
+
+
+    def callback_conn(self, ch_index, value):
+        """
+        Callback executed when a channel connection status is changed.
+
+        Parameters
+        ----------
+        ch_index : int
+            The channel index on the list for this rule.
+        value : bool
+            Whether or not this channel is connected.
+
+        """
+        self._connections[ch_index] = value
+        self._send_update(self.connected, self._value)
+
+    def warn_unconnected_channels(self):
+        logger.debug(
+            "Calculation '%s': Not all channels are connected, skipping execution.",
+            self.objectName())
+
+    def calculate_expression(self):
+        """
+        Evaluate the expression defined by the rule and emit the `rule_signal`
+        with the new value.
+        """
+        vals = list(self._values)
+
+        if None in vals:
+            logger.debug('Skipping execution as not all values are set.')
+            return
+
+        eval_env = {'math': math,
+                    'np': np,
+                    'ch': vals}
+        eval_env.update({k: v
+                         for k, v in math.__dict__.items()
+                         if k[0] != '_'})
+
+        try:
+            ret = eval(self._expression, eval_env)
+            self._value = ret
+            self._send_update(self.connected, ret)
+        except Exception as e:
+            logger.exception("Error while evaluating CalcPlugin connection %s",
```

Comment:
Should repeated failures cause this to stop updating or throttle updates somehow?